### PR TITLE
Fixed focus behavior for click automation (closes #770)

### LINKED ISSUE #770: Investigate and extend focus behavior in click automation
At the current moment in all browsers in some cases we call focus after mousedown on non-focusable elements:
https://github.com/DevExpress/testcafe/blob/master/src/client/automation/utils/utils.js#L43
https://github.com/DevExpress/testcafe/blob/master/src/client/automation/utils/utils.js#L53
It was expected that it is the reason of failed ASPx tests in IE, but while we were fixing this, was found, that IE has different focus logic and it's the reason of failed ASPx tests. So, we should fix logic over non-focusable elements in all browsers and change focus behavior in IE.

### PR DESCRIPTION
**Pull request by georgiy-abbasov**
\cc @inikulin, @AlexanderMoskovkin, @helen-dikareva

### CODE REVIEW

File: src/client/automation/playback/click/index.js [hunk 1]
```diff
@@ -160,6 +160,8 @@ export default class ClickAutomation {
                 // NOTE: in IE focus is not raised if element was focused
                 // before click, even if focus is lost during mousedown
                 this.eventState.simulateFocus = !browserUtils.isIE || activeElement !== this.eventArgs.element ||
+                                                !domUtils.isElementFocusable(this.eventArgs.element) &&
```
Comment (helen-dikareva) [inline src/client/automation/playback/click/index.js]: Let's move this check in utils. Is it possible?

File: test/functional/fixtures/api/es-next/click/testcafe-fixtures/click-test.js [hunk 2]
```diff
@@ -84,3 +84,11 @@ test('Selector returns text node', async t => {
 
     await t.click(getNode);
 });
+
+const getActiveElementTagName = ClientFunction(() => window.document.activeElement.tagName.toLowerCase());
```
Comment (helen-dikareva) [inline test/functional/fixtures/api/es-next/click/testcafe-fixtures/click-test.js]: We just can compare `document.activeElement` and `document.body`

File: test/functional/fixtures/regression/gh-770/testcafe-fixtures/index.test.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { ClientFunction } from 'testcafe';
+import { expect } from 'chai';
+
+
+fixture `GH-770`
+    .page `http://localhost:3000/fixtures/regression/gh-770/pages/index.html`;
+
+const getActiveElementTagName = ClientFunction(() => window.document.activeElement.tagName.toLowerCase());
```
Comment (helen-dikareva) [inline test/functional/fixtures/regression/gh-770/testcafe-fixtures/index.test.js]: The same as above

File: src/client/automation/playback/click/index.js (hunk 1 above)
Comment (georgiy-abbasov) [inline src/client/automation/playback/click/index.js]: Unfortunately no. Active element can change after performing mousedown.

Comment (AlexanderMoskovkin) [inline src/client/automation/playback/click/index.js]: Why?
```js
function needSimulateFocusInIE (element) {
    var activeElement = domUtils.getActiveElement();

    return activeElement !== element || !domUtils.isElementFocusable(element) && domUtils.isBodyElement(activeElement) || isContentEditableBody;
}
```

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
FPR

File: src/client/automation/playback/click/index.js [hunk 4]
```diff
@@ -150,16 +151,13 @@ export default class ClickAutomation {
...
-                var isContentEditableBody = isBodyElement && isContentEditable;
                 var activeElement         = domUtils.getActiveElement();
 
+                this.beforeMouseDownActiveElement = activeElement;
+
                 // NOTE: in IE focus is not raised if element was focused
                 // before click, even if focus is lost during mousedown
-                this.eventState.simulateFocus = !browserUtils.isIE || activeElement !== this.eventArgs.element ||
-                                                isContentEditableBody;
+                this.eventState.simulateFocus = !browserUtils.isIE;
```
Comment (helen-dikareva) [inline src/client/automation/playback/click/index.js]: If now `this.eventState.simulateFocus` is assigned only in `focus` method for IE https://github.com/DevExpress/testcafe/pull/843/files#diff-0b8d22328e7dddcc102e65786aa60657R224
we can move this part of code there too

File: src/client/automation/utils/utils.js [hunk 5]
```diff
@@ -40,6 +40,7 @@ export function focusAndSetSelection (element, simulateFocus, caretPos) {
         var activeElement               = domUtils.getActiveElement();
         var isTextEditable              = domUtils.isTextEditableElement(element);
         var labelWithForAttr            = domUtils.closest(element, 'label[for]');
+        var isElementIsFocusable        = domUtils.isElementFocusable(element);
```
Comment (helen-dikareva) [inline src/client/automation/utils/utils.js]: may be just `isElementFocusable`

File: src/client/automation/utils/utils.js [hunk 6]
```diff
@@ -57,6 +58,16 @@ export function focusAndSetSelection (element, simulateFocus, caretPos) {
             resolve();
             return;
         }
+        else if (!isElementIsFocusable && !isContentEditable) {
```
Comment (helen-dikareva) [inline src/client/automation/utils/utils.js]: We don't need `else` because previous `if` has `return` in any cases

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
FPR

**Comment by georgiy-abbasov** [comment]
ping @VasilyStrelyaev

**Comment by georgiy-abbasov** [comment]
ping @inikulin

File: src/client/automation/playback/click/index.js [hunk 7]
```diff
@@ -40,8 +40,9 @@ export default class ClickAutomation {
         this.offsetX = clickOptions.offsetX;
         this.offsetY = clickOptions.offsetY;
 
-        this.targetElementParentNodes = [];
-        this.clickEventElement        = null;
+        this.targetElementParentNodes     = [];
+        this.clickEventElement            = null;
+        this.beforeMouseDownActiveElement = null;
```
Comment (inikulin) [inline src/client/automation/playback/click/index.js]: `beforeMouseDownActiveElement` - I can't understand meaning of this var due to bad wording. Is it `activeElementBeforeMouseDown`?

File: test/functional/fixtures/api/es-next/click/test.js [hunk 8]
```diff
@@ -123,4 +123,8 @@ describe('[API] t.click()', function () {
                 });
         });
     });
+
```
Comment (inikulin) [inline test/functional/fixtures/api/es-next/click/test.js]: Why one test is in regression section and the other one is in click? That's the logic behind it?

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

File: test/functional/fixtures/api/es-next/click/test.js (hunk 8 above)
Comment (georgiy-abbasov) [inline test/functional/fixtures/api/es-next/click/test.js]: My fault, i've placed them correctly for now.
I've missed some details in issue description, so i changed it to be clear, what we fix.

**Comment by georgiy-abbasov** [comment]
FPR

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by georgiy-abbasov** [comment]
Cause @inikulin isn't around, \cc @churkin, review please

**Comment by helen-dikareva** [comment]
You need to remove `simulateFocus` from here https://github.com/DevExpress/testcafe/blob/master/src/client/automation/playback/click/index.js#L53

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Pull request merged**